CG display lists stick around on layers even after backing store is removed

hortont424 opened the pull request:
#### 545623ba3da1f4867c26ed80feefec663f0e45c2
<pre>
CG display lists stick around on layers even after backing store is removed
<a href="https://bugs.webkit.org/show_bug.cgi?id=252777">https://bugs.webkit.org/show_bug.cgi?id=252777</a>
rdar://104935890

Reviewed by Dean Jackson.

When removing backing store from a layer (because e.g. it became `visibility: hidden`),
we need to also remove the CG display list blob.

* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(-[CALayer _web_clearContents]):
Factor out &quot;clear contents and display list if installed&quot; code.

Work around a bug where it&apos;s currently not possible to reset the display list to nil
without causing the layer to create its own backing store, by assigning an empty display list instead.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::applyBackingStoreToLayer):
If we don&apos;t have layer contents for some unexpected reason (e.g. IPC timeout),
clear both contents and display list.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
If we don&apos;t have backing store at all, clear both contents and display list.

Canonical link: <a href="https://commits.webkit.org/260720@main">https://commits.webkit.org/260720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57e786696a4ca273eb3856580d4d5021ae393061

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18246 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/693 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113049 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19705 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9548 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101407 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114925 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29654 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11024 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/30997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50600 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7395 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13372 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
```diff
@@ -119,6 +123,20 @@ - (BOOL)_web_maskMayIntersectRect:(CGRect)rect
+    if ([self valueForKeyPath:WKCGDisplayListContentsKey]) {
+        [self setValue:nil forKeyPath:WKCGDisplayListContentsKey];
+        [self setValue:nil forKeyPath:WKCGDisplayListPortsKey];
+        [self setNeedsDisplay];
```
This is a bit odd. Often you clear contents to remove backing store, but calling `-setNeedsDisplay` will trigger a `-display` which will reallocate backing store.

hortont424 replied:
good... point

hortont424 replied:
Can't remove the setNeedsDisplay yet, but added a workaround that will prevent backing store from being allocated for now.

hortont424 committed:
CG display lists stick around on layers even after backing store is removed
https://bugs.webkit.org/show_bug.cgi?id=252777
rdar://104935890

Reviewed by Dean Jackson.

When removing backing store from a layer (because e.g. it became `visibility: hidden`),
we need to also remove the CG display list blob.

* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(-[CALayer _web_clearContents]):
Factor out "clear contents and display list if installed" code.

Work around a bug where it's currently not possible to reset the display list to nil
without causing the layer to create its own backing store, by assigning an empty display list instead.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::applyBackingStoreToLayer):
If we don't have layer contents for some unexpected reason (e.g. IPC timeout),
clear both contents and display list.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
If we don't have backing store at all, clear both contents and display list.

Canonical link: https://commits.webkit.org/260720@main
